Std/Classes/LawfulAlternative.lean

```diff
@@ -0,0 +1,68 @@
+import Std.Classes.LawfulMonad
+import Lean.Linter
+
+/-- Combination class of `Monad` and `Alternative` with a common `Applicative` component -/
+class MonadAlternative (m : Type _ → Type _) extends Monad m, Alternative m
+
+-- Missing docs in core
+attribute [nolint docBlame] MonadAlternative.failure MonadAlternative.orElse
+
+/-- Basic laws for `Alternative` -/
+class LawfulAlternative (f : Type _ → Type _) [Alternative f] extends LawfulApplicative f : Prop where
+  /-- Map failure is failure -/
+  map_failure (g : α → β) : g <$> (failure : f α) = failure
+  /-- Map distributes over `orElse` -/
+  map_orElse (g : α → b) (x y : f α) : g <$> (x <|> y) = (g <$> x <|> g <$> y)
+  /-- Failure is a left identity for `orElse`  -/
+  failure_orElse (x : f α) : (failure <|> x) = x
+  /-- Failure is a right identity for `orElse` -/
+  orElse_failure (x : f α) : (x <|> failure) = x
+  /-- `orElse` is associative -/
+  orElse_assoc (x y z : f α) : ((x <|> y) <|> z) = (x <|> (y <|> z))
+
+export LawfulAlternative (map_failure map_orElse failure_orElse orElse_failure orElse_assoc)
+
+attribute [simp] map_failure failure_orElse
+
+instance : LawfulAlternative Option where
+  map_failure        := by intros; rfl
+  map_orElse _ x _   := by cases x <;> rfl
+  failure_orElse     := by intros; rfl
+  orElse_failure x   := by cases x <;> rfl
+  orElse_assoc x _ _ := by cases x <;> rfl
+
+namespace ReaderT
+
+@[simp] theorem run_failure [Monad m] [Alternative m] (ctx : ρ) : (failure : ReaderT ρ m α).run ctx = failure := rfl
+
+@[simp] theorem run_orElse [Monad m] [Alternative m] (x y : ReaderT ρ m α) (ctx : ρ) : (x <|> y).run ctx = (x.run ctx <|> y.run ctx) := rfl
+
```

Comment:
I'd prefer these (and some declarations below that don't mention `LawfulAlternative` find files higher up to live in. (Perhaps even in the `Lean` directory.)

Reply:
Agreed. (This PR is currently low priority for me so it might take me a while to get back to this one. I eventually will though!)